dbms/src/Interpreters/Aggregator.cpp

```diff
@@ -2572,6 +2139,154 @@ void Aggregator::setCancellationHook(CancellationHook cancellation_hook)
+        aggregator.mergeSingleLevelDataImpl<AggregationMethodName(NAME)>(data); \
+        break;                                                                  \
+    }
+        switch (first->type)
+        {
+            APPLY_FOR_VARIANTS_SINGLE_LEVEL(M)
+        default:
+            throw Exception("Unknown aggregated data variant.", ErrorCodes::UNKNOWN_AGGREGATED_DATA_VARIANT);
+        }
+#undef M
+        single_level_blocks = aggregator.prepareBlocksAndFillSingleLevel(*first, final);
+    }
+    ++current_bucket_num;
+    return popBlocksListFront(single_level_blocks);
+}
+
+Block MergingBuckets::getDataForTwoLevel(size_t concurrency_index)
+{
+    assert(concurrency_index < two_level_parallel_merge_data.size());
+    auto & two_level_merge_data = *two_level_parallel_merge_data[concurrency_index];
+
+    Block out_block = popBlocksListFront(two_level_merge_data);
+    if (likely(out_block))
+        return out_block;
+
+    if (current_bucket_num >= NUM_BUCKETS)
+        return {};
+    while (true)
+    {
+        auto local_current_bucket_num = current_bucket_num.fetch_add(1);
+        if (unlikely(local_current_bucket_num >= NUM_BUCKETS))
+            return {};
+
+        doLevelMerge(local_current_bucket_num, concurrency_index);
+        Block out_block = popBlocksListFront(two_level_merge_data);
+        if (likely(out_block))
+            return out_block;
+    }
+}
+
+void MergingBuckets::doLevelMerge(Int32 bucket_num, size_t concurrency_index)
+{
+    auto & two_level_merge_data = *two_level_parallel_merge_data[concurrency_index];
+    assert(two_level_merge_data.empty());
+
+    assert(!data.empty());
+    auto & merged_data = *data[0];
+    auto method = merged_data.type;
+
+    /// Select Arena to avoid race conditions
+    Arena * arena = merged_data.aggregates_pools.at(concurrency_index).get();
+
+#define M(NAME)                                                                           \
+    case AggregationMethodType(NAME):                                                     \
+    {                                                                                     \
+        aggregator.mergeBucketImpl<AggregationMethodName(NAME)>(data, bucket_num, arena); \
+        two_level_merge_data = aggregator.convertOneBucketToBlocks(                       \
+            merged_data,                                                                  \
+            *ToAggregationMethodPtr(NAME, merged_data.aggregation_method_impl),           \
+            arena,                                                                        \
+            final,                                                                        \
+            bucket_num);                                                                  \
+        break;                                                                            \
+    }
+    switch (method)
+    {
+        APPLY_FOR_VARIANTS_TWO_LEVEL(M)
+    default:
+        break;
```

Comment:
Why not throw error for the default branch?

Reply:
ok, updated.